### ISSUE TITLE
[ENH] Add optional TC and weight drift in `cross_val_predict`

### ISSUE THREAD

**Issue by HugoDelatte** [skfolio/skfolio#161]
* Add optional  TC and  weight drift in `cross_val_predict`
* Add a doc section that explain the pros/cons for using non-compounded vs compounded returns / no weight drift vs  weight drift / smooth TC vs exact TC --> in term of unbiased expectation of cumulative simple return (e.g. measuring strategy "skill" or “alpha”) vs expected terminal wealth vs  single sample (historical) path.

**Comment by HugoDelatte** [comment]
We also need to ensure that in `cross_val_predict`, when the optimization requires previous_weights (for TC, turnovers, etc.), we disable parallelization so the portfolio is computed sequentially, passing the weights from one run to the next.

**Comment by moyuanchen** [comment]
Took a crack at this. I added an optional transaction_costs parameter to cross_val_predict and if it is not none the method sequentially fit the walk forward cross validation taking into account the weight drift and transaction cost.

**Comment by moyuanchen** [comment]
### Weight-drift
The change in the price of the stocks in a portfolio results in a shift in their weights.

$$
w_s = w_s(t)
$$
$$
w_s(0)=w_{s,0} 
$$

Where $w_{s,0}$ denote the initial weight given to each stock in a portfolio. Normalizing the total initial portfolio value to 1, the weight is equal to the value of the shares in the stock. Define simple return $r_s(t) = p_s(t)/p_s(0)$, where $p_s(t)$ is the price of the stock at time $t$. We have

$$
w_s(t) = w_{s,0}r_s(t); \quad dw_s = w_{s,0}dr_s
$$

In a simple one-period buy and hold portfolio, the weight drift is trivial. The cumulative return of the portfolio is

$$
r_P(t) =\frac{ \sum_{s\in\mathcal{S}}w_s(t)}{\sum_{s\in\mathcal{S}}w_{s,0}}= \sum_{s\in\mathcal{S}}  w_{s,0}r_s(t)
$$
### Discretization
Consider a NxP matrix $X$ where N is the total number of observations and P the number of stocks in the portfolio. 
The return of the portfolio is calculated by directly multiplying the simple return of the stock with its weight. The $t^{th}$ row is the simple return of each stock in that period $X_{t,s} = p_{t,s} / p_{t-1,s} $. The portfolio wide return of period t can be expressed as

$$
\tilde{r}_t = X_t\cdot w_0
$$

Where $w_0$ is a P length vector of the weights generated from the optimization. Note that this is no longer the cumulative return but the return of period t. It is also important to note that the actual weights of the stocks would have drifted at time $t$ based on its previous returns.
### Compounded vs Non-compounded Return
The cumulative non-compounded return is the sum of the period returns up to time t.


$$
\sum_{t'=1}^{t} \tilde{r_{t'}} =  \sum_{t'=1}^{t} \sum_{s\in\mathcal{S}} X_{t', s} w_{0,s}= \sum_{s\in\mathcal{S}}w_{0,s} \sum_{t'=1}^{t} X_{t', s}
$$

The effect of weight drift shows up in the non-compounded return calculated this way because

$$
\sum_{t'=1}^{t} X_{t', s} = \sum_{t'=1}^{t} \frac{ p_{t',s} }{ p_{t'-1,s} }
$$

and

$$
r_s(t) = \frac{ p_s(t)}{p_s(0)} = \prod_{t'=1}^{t} \frac{ p_{t',s} }{ p_{t'-1,s} }
$$

### Transaction Costs
As outlined in the example: https://skfolio.org/auto_examples/mean_risk/plot_6_transaction_costs.html, using the walk-forward cross validation method the expected transaction cost can be added to the optimization problem. This way the transaction cost is no longer zero, and the weight drift in the non-compounded return is reset when the portfolio is rebalanced.

Consider the optimization problem

$$
argmin_w \sum_{s,s' \in \mathcal{S}}^n w_s \sigma_{s,s'} w_s'
$$
with constrains
$$
\sum_{s\in \mathcal{S}} w_s \lt r_s \gt =r_{target}
$$
$$
\sum_{s\in \mathcal{S}} w_s = 1
$$

where the brackets denotes expected value, which can be calculated from arithmetic or geometric mean, corresponding to non-compounded and compounded return. Consider the expected return with transaction cost taken into account with the exact or smooth transaction cost

$$
\lt r_s \gt = w_s \mu_s -c_s |w_s - w_{prev}|
$$
$$
\lt r_s \gt = w_s \mu_s -c_s \sqrt{(w_s - w_{prev})^2 + \epsilon}
$$
### Pros and Cons
The user can choose which one of weight drift, compounding, and smooth transaction costs to turn on or off. In the following section we discuss how these options affect the outcome of the optimization and how should one choose depending on their goals.
#### Arithmetic vs. Geometric Mean
The geometric mean represents the long-term average return of the portfolio/stock. On the other hand, the arithmetic mean is the center of the return distribution in the training period.  The arithmetic mean is the better future expectation value of returns for portfolio optimization, while the geometric mean captures the return over the whole period and can be used for historical backtesting.
#### Smoothed transaction cost
When optimizing a portfolio the smoothed transaction cost is more useful because it is differentiable and works with the solver, at the cost of introducing a small bias $\epsilon$.

To be continued...

**Comment by moyuanchen** [comment]
Thank you @CarloNicolini and @HugoDelatte for your feedback! 

@CarloNicolini Online portfolio selection sounds very interesting! If we define regret as for example

$$
R_T = \sum_t^T max(\\{ X_{t,s} | s\in\mathcal{S} \\}) - \sum_t^T \tilde{r}_t
$$

We can use this as the loss function and use a `partial_fit` and online gradient descent to find some optimal model that minimizes this quantity; however, this is no longer minimizing the variance.

To minimize variance perhaps we need some kind of loss function on the variance of the portfolio?

Looking forward to further discussions!

**Comment by moyuanchen** [comment]
Picking the weight-drift half of this back up.

In `Portfolio.__init__` (`src/skfolio/portfolio/_portfolio.py:590`):

```python
returns = weights @ rets.T - total_cost - total_fee
```

The dot product implies **rebalancing back to `w` at every observation**. Adding a flag here enables user to backtest a buy-and-hold strategy.

Illustrating the point with two assets at 50/50, three periods of +10% / −10%:

```python
import numpy as np
from skfolio import Portfolio

X = np.array([[0.10, -0.10], [0.10, -0.10], [0.10, -0.10]])
w0 = np.array([0.5, 0.5])

# skfolio: constant weights, re-levered every period
print("rebalancing    :", Portfolio(X=X, weights=w0).returns)

# buy-and-hold: weights drift with the assets
v = w0.copy()
values = []
for r in X:
    v = v * (1 + r)
    values.append(v.sum())
values = np.array(values)
prev = np.r_[1.0, values[:-1]]
print("buy-and-hold       :", (values - prev) / prev)
print("end drifted weights:", v / v.sum())
```

```
rebalancing    : [0. 0. 0.]
buy-and-hold       : [0.         0.01       0.01980198]
end drifted weights: [0.6461165 0.3538835]
```

This also decides what carries across folds. `model_selection/_validation.py:538`:

```python
prev_weights = ptf.weights_dict if use_dict else ptf.weights
```

hands the *target* weights to the next fold. If the user declared buy-and-hold, it should hand the end-of-fold drifted weights, so turnover and TC are charged against where the book actually ended up.

## Question

**Should the flag live on `cross_val_predict`?** I think yes. It is a statement about the strategy being backtested and the walk-forward is where the user expresses that. Alternatively we can put it in Portfolio, and add a property that returns drifted or plain weights depending on the flag. Also what should we call the flag (`weight_drift`, `buy_and_hold`, `rebalancing`)?


## Docs placement

For the pros/cons section from your original note:

- `docs/user_guide/portfolio.rst:107-127` — the existing warning block already covers non-compounded vs compounded; the rebalance-vs-hold axis extends it.
- `docs/user_guide/model_selection.rst:27` (*Cross-Validation Prediction*) — what a test fold represents under each convention, plus the cross-fold `previous_weights` handoff.
- `examples/mean_risk/plot_6_transaction_costs.py` — the same walk-forward under both conventions, once the flag exists.
- `Portfolio` docstring (`_portfolio.py:38-55`) mirrors the user-guide warning and would need to stay in sync.

**Comment by CarloNicolini** [comment]
This point has already been discussed with Hugo.
Using weight drift breaks many homogeneity assumptions in the optimizers.
The library would change a lot.

I’d suggest you to take a look at my branch universal_portfolio that
implements online portfolio selection methods where the weight drift
naturally enters the optimization process.

https://github.com/CarloNicolini/skfolio/tree/universal_portfolio

I already had some back-testers using weight drift though, but the scope of
the library would change a lot. One of my earliest implementations is
present in scikit-portfolio. I could point it to you. It was heavily
inspired to the one from the matlab financial toolbox.


Il giorno dom 23 ago 2026 alle 00:22 moyuanchen ***@***.***>
ha scritto:

**Comment by HugoDelatte** [comment]
Thank you @CarloNicolini and @moyuanchen for brainstorming on this.
After reflection, I think there is an easy API way to let the user chose, the same way we do with `compounded`.
Let me come back with more details on this.
